sdk-typed-search-attributes.md

```diff
@@ -0,0 +1,497 @@
+  * Search attributes setting on start - via `Record<string, scalar[]>` on start options
+  * Search attributes getting on list/describe - as `Record<string, scalar[]>`
+* Python
+  * Search attributes getting in workflow - on info as `Mapping[str, List[scalar]]`
+  * Search attributes setting in workflow - via `upsert_search_attributes` accepting `Mapping[str, List[scalar]]`
+  * Search attributes setting on start - via `Mapping[str, List[scalar]]` on start options
+  * Search attributes getting on list/describe - as `Mapping[str, List[scalar]]`
+* .NET
+  * Still in alpha, can decide whatever we want
+
+### Server
+
+* ElasticSearch
+  * Multi-value search attribute for non-keyword-list - accepted, but discouraged, and returned as given
+    * TODO(cretz): Question outstanding to server team about the benefits of multi-value non-keyword-list attributes
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+* SQL
+  * Multi-value search attribute for non-keyword-list - only allowed if a single value, always returned single value
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+
+### Problems with current behavior
+
+* Go and Java rely on user to resolve ambiguity between single and multi value search attributes (Go is worse because it has no helpers)
+* TypeScript and Python always disambiguate making all search attributes a list (and Java too in single-value cases) even if list not allowed anymore
+* Setting invalid search attributes in workflow never returns a failure that can be handled by workflow code
+* Users have no way to clarify the difference between keyword and keyword-list
+
+## Goals
+
+Immediate:
+
+* As much as is supported by the language, users need to be able to get and set individually typed search attributes
+* Maintain backwards compatibility but deprecate any approach that doesn't let you identify the search attribute type
+  alongside the use of its value
+
+Non-immediate:
+
+* Provide ways to define predefined search attributes that are typed
+* Also provide ways to define predefined memo values that are typed
+
+Non-goals:
+
+* Better validation of actual search attribute type from server
+* Better way for server to tell SDK the upsert failed
+* Support for user-explicit upsert of multiple typed search attributes in one statement
+* Support for typed registration of search attributes
+* Support for predefined search attributes in any form of query building
+
+## Proposed solution
+
+### General
+
+* All SDK workflow APIs will now have a mutable search attribute map with typed getters and setters
```

Comment:
Not at all a problem with this proposal - but we will want to be clear in docs/docstrings that the search attribute map you see in the workflow may not necessarily contain a complete representation of actual search attributes in persistence if users ever modify them outside of a workflow.

Reply:
We have an event for external workflow SA mod now right? I think the only way they can mod them outside of workflow is with that which will be put into history. And if there's race, the workflow will get an unhandled command and retry. I sure hope the mutate-search-attribute API if/when it gets created has some form of conflict token if it loses the race.

The only way that this may not be accurate I think is if there is a server validation error with the search attribute (since we don't get that feedback).

Reply:
Will bring up in discussion. Lots of stuff in here is invalid if we need to error on invalid search attribute setting, so we can decide to invalidate here in proposal or invalidate later in code.

Reply:
I don't think we want to error on failure here, I think that makes a worse API and I don't see much benefit to it.
How would a user be expected to handle upsert failure? We might as well just fail the workflow task like we do today.

Reply:
> How would a user be expected to handle upsert failure?

Swallow it because I don't care whether the search attribute is present on the server.

> let user deploy a fixed version of the workflow or define the missing search attributes.

This will become a much less viable option now that we restrict search attribute count to a small number.

Note, I don't mind never erroring, I just want to present all sides here.